gson/src/main/java/com/google/gson/stream/JsonReader.java

```diff
@@ -1255,22 +1262,45 @@ public void skipValue() throws IOException {
         stackSize--;
         count--;
       } else if (p == PEEKED_END_OBJECT) {
+        // Only update when object end is explicitly skipped, otherwise stack is not updated anyways
+        if (count == 0) {
+          pathNames[stackSize - 1] = null; // Free the last path name so that it can be garbage collected
+        }
         stackSize--;
         count--;
-      } else if (p == PEEKED_UNQUOTED_NAME || p == PEEKED_UNQUOTED) {
+      } else if (p == PEEKED_UNQUOTED) {
```

Comment:
While you are here, could you change this to a `switch` statement too? Maybe the JIT compiler is smart enough to avoid having to do a sequence of up to 12 comparisons, but we could avoid relying on that.